building_sim_plugins/building_ignition_plugins/src/crowd_simulator.cpp

```diff
@@ -76,192 +80,167 @@ void CrowdSimulatorPlugin::PreUpdate(
     {
-      auto it_objects_name = objects_name.find(name->Data());
-      if (it_objects_name != objects_name.end())
+      objects_name.insert({crowd_sim_interface->get_internal_agent(id), id});
+    }
+    // for internal agent
+    ecm.Each<ignition::gazebo::components::Actor,
+      ignition::gazebo::components::Name>(
+      [&](const ignition::gazebo::Entity& entity,
+      const ignition::gazebo::components::Actor* actor,
+      const ignition::gazebo::components::Name* name) -> bool
       {
-        // update in entityDic
-        _entity_dic[name->Data()] = entity;
-        auto obj_ptr =
-        _crowd_sim_interface->get_object_by_id(it_objects_name->second);
-        // config internal spawned agent for custom trajectory
-        if (!obj_ptr->is_external)
+        auto it_objects_name = objects_name.find(name->Data());
+        if (it_objects_name != objects_name.end())
         {
-          _config_spawned_agents(obj_ptr, entity, ecm);
+          // update in entityDic
+          _entity_dic[name->Data()] = entity;
+          // config internal spawned agent for custom trajectory
+          _config_spawned_agents(nullptr, entity, ecm, crowd_sim_interface);
+          objects_name.erase(name->Data());
         }
-        objects_name.erase(name->Data());
-        RCLCPP_INFO(_crowd_sim_interface->logger(),
-        "Crowd Simulator found agent: " + name->Data() );
-      }
-      return true;
+        return true;
+      });
+    if (objects_name.size() != 0)
+    {
+      _initialized = false;
+      return;
     }
-    );
-  // for internal agent
-  ecm.Each<ignition::gazebo::components::Actor,
-    ignition::gazebo::components::Name>(
-    [&](const ignition::gazebo::Entity& entity,
-    const ignition::gazebo::components::Actor*,
-    const ignition::gazebo::components::Name* name) -> bool
+    _initialized = true;
+  }
+  else
+  {
+    for (size_t id = 0; id < object_count; id++)
     {
-      auto it_objects_name = objects_name.find(name->Data());
-      if (it_objects_name != objects_name.end())
+      auto obj = crowd_sim_interface->get_object_by_id(id);
+      // already found in the Dic
+      if (_entity_dic.find(obj->model_name) != _entity_dic.end())
+        continue;
+      objects_name.insert({obj->model_name, id});
+    }
+    // for external agent
+    ecm.Each<ignition::gazebo::components::Model,
+      ignition::gazebo::components::Name>(
+      [&](const ignition::gazebo::Entity& entity,
+      const ignition::gazebo::components::Model*,
+      const ignition::gazebo::components::Name* name) -> bool
       {
-        // update in entityDic
-        _entity_dic[name->Data()] = entity;
-        auto obj_ptr =
-        _crowd_sim_interface->get_object_by_id(it_objects_name->second);
-        // config internal spawned agent for custom trajectory
-        if (!obj_ptr->is_external)
+        auto it_objects_name = objects_name.find(name->Data());
+        if (it_objects_name != objects_name.end())
```

Comment:
I noticed that for `ecm.Each`, it filters out actors with name instances (including actors that might not concern crowd_sim - might happen in future maps). The `objects_name` keeps a internal records (name and id) of entities that are relevant to crowdsim and only operate on them, with `Object` structure that contains essential information (referenced by id).
Currently, I hardly see any use for `entity_dic` other than keeping a record, and does nothing else but some additional checking (to be confirmed again).
Anyway, these are legacy components.

Reply:
I found the use for `entity_dic`. For ignition, `entity_dic` store the internal and external agents ecm entities so that they an be accessed directly for their current position in the world. Then, their current position would be updated in Menge (for external agents especially) for processing.